src/python/pants/core/util_rules/gen_rule.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+
+from pants.core.target_types import (
+    FilesSources,
+    GenRuleCommandField,
+    GenRuleOutputsField,
+    GenRuleSources,
+    GenRuleToolsField,
+)
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import (
+    EMPTY_SNAPSHOT,
+    AddPrefix,
+    CreateDigest,
+    Digest,
+    Directory,
+    GlobMatchErrorBehavior,
+    MergeDigests,
+    PathGlobs,
+    Snapshot,
+)
+from pants.engine.process import (
+    BashBinary,
+    BinaryNotFoundError,
+    BinaryPathRequest,
+    BinaryPaths,
+    Process,
+    ProcessResult,
+    SearchPath,
+)
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    GeneratedSources,
+    GenerateSourcesRequest,
+    Sources,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+
+
+class GenerateFilesFromGenRuleRequest(GenerateSourcesRequest):
+    input = GenRuleSources
+    output = FilesSources
+
+
+@rule(desc="Running gen_rule", level=LogLevel.DEBUG)
+async def run_gen_rule(
+    request: GenerateFilesFromGenRuleRequest, bash: BashBinary
+) -> GeneratedSources:
+    gen_rule = request.protocol_target
+    working_directory = gen_rule.address.spec_path
+    command = gen_rule[GenRuleCommandField].value
+    tools = gen_rule[GenRuleToolsField].value
+    outputs = gen_rule[GenRuleOutputsField].value
+    source_globs = gen_rule[GenRuleSources].value or ()
+
+    if not (command and tools and outputs):
+        return GeneratedSources(EMPTY_SNAPSHOT)
+
+    tool_requests = [
+        BinaryPathRequest(
+            binary_name=tool,
+            search_path=SearchPath(("/usr/bin", "/bin", "/usr/local/bin")),
+        )
+        for tool in tools
+    ]
+    tool_paths = await MultiGet(
+        Get(BinaryPaths, BinaryPathRequest, request) for request in tool_requests
+    )
+
+    tools_env: dict[str, str] = {}
+    for binary, tool_request in zip(tool_paths, tool_requests):
+        if binary.first_path:
+            tools_env[tool_request.binary_name] = binary.first_path.path
+        else:
+            raise BinaryNotFoundError(
+                tool_request, rationale=f"execute gen_rule {gen_rule.address}"
+            )
+
+    transitive_targets = await Get(
+        TransitiveTargets,
+        TransitiveTargetsRequest([gen_rule.address]),
+    )
+
+    own_sources, dep_sources = await MultiGet(
+        Get(
+            Digest,
+            PathGlobs(
+                [os.path.join(working_directory, glob) for glob in source_globs],
+                glob_match_error_behavior=GlobMatchErrorBehavior.error,
+                description_of_origin=f"{gen_rule.address}'s `{GenRuleSources.alias}` field",
+            ),
```

Comment:
You can call [Sources.path_globs](https://github.com/pantsbuild/pants/blob/9c12fc656813d8938342b115cc8c445418c6d93e/src/python/pants/engine/target.py#L1431-L1451) for this.

Reply:
Ah, thanks :)